### PR TITLE
Add E0088, E0091, E0109 and E0110 and error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #24407.
cc @michaelsproul

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -934,6 +934,28 @@ The number of supplied parameters much exactly match the number of defined type
...
+
+```
+fn f() {}
+
+fn main() {
+    f::<'static>() // error: too many lifetime parameters provided
+}
+```
+
+Please check you give the right number of lifetime parameter. Example:
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: parameters

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -934,6 +934,28 @@ The number of supplied parameters much exactly match the number of defined type
 parameters.
 "##,
 
+E0088: r##"
+You gave too many lifetimes parameters. Erroneous code example:
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: lifetime parameters

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -934,6 +934,28 @@ The number of supplied parameters much exactly match the number of defined type
...
+
+```
+fn f() {}
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: I'd like some more functions here with two or three lifetimes as well.

File: src/librustc_typeck/diagnostics.rs [hunk 4]
```diff
@@ -959,6 +981,20 @@ fn main() {
 ```
 "##,
 
+E0091: r##"
+You gave an unnecessary type parameter. Erroneous code example:
```
Comment (michaelsproul) [inline src/librustc_typeck/diagnostics.rs]: I think it's worth noting the specific context:

> You provided an unnecessary type parameter in a type alias.

File: src/librustc_typeck/diagnostics.rs [hunk 5]
```diff
@@ -959,6 +981,20 @@ fn main() {
...
+
+```
+type Foo<T> = u32; // error: type parameter `T` is unused
+```
+
+Please check you didn't write to many type parameter. Example:
```
Comment (michaelsproul) [inline src/librustc_typeck/diagnostics.rs]: "to" -> "too"

Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: type parameters

File: src/librustc_typeck/diagnostics.rs [hunk 6]
```diff
@@ -959,6 +981,20 @@ fn main() {
 ```
 "##,
 
+E0091: r##"
+You gave an unnecessary type parameter. Erroneous code example:
+
+```
+type Foo<T> = u32; // error: type parameter `T` is unused
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: I'd like some examples with type parameters on the RHS too.

File: src/librustc_typeck/diagnostics.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I'll add this note. However, I didn't find out yet out to "force pass" lifetime parameter to a function.

File: src/librustc_typeck/diagnostics.rs (hunk 6 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: RHS ?

File: src/librustc_typeck/diagnostics.rs (hunk 3 above)
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: It works in the scenario when you're in a surrounding function or impl block which has defined lifetimes.

File: src/librustc_typeck/diagnostics.rs (hunk 6 above)
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: right hand side

File: src/librustc_typeck/diagnostics.rs [hunk 7]
```diff
@@ -959,6 +1004,21 @@ fn main() {
 ```
 "##,
 
+E0091: r##"
+You gave an unnecessary type parameterin a type alias. Erroneous code
```
Comment (apasel422) [inline src/librustc_typeck/diagnostics.rs]: s/parameterin/parameter in/

File: src/librustc_typeck/diagnostics.rs (hunk 3 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: I might need some help. I don't understand what you mean.

File: src/librustc_typeck/diagnostics.rs (hunk 6 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: You mean like this:

```Rust
type X = u32<i32>;
```

? If yes, it gives E0109 (which I added to this PR).

Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: No, I meant something like `type X<A,B> = Box<A>`.

File: src/librustc/diagnostics.rs [hunk 8]
```diff
@@ -360,6 +360,39 @@ integer type:
...
 
+E0109: r##"
+You tried to give type parameter to a type which doesn't need it. Erroneous
+code example:
+
+```
+type X = u32<i32>; // error: type parameters are not allowed on this type
+```
+
+Please check you actually used the good type or check again its definition.
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: "Please check that you used the correct type and recheck its definition. Perhaps it doesn't need the type parameter."

File: src/librustc_typeck/diagnostics.rs (hunk 6 above)
Comment (GuillaumeGomez) [inline src/librustc_typeck/diagnostics.rs]: Okay ! My bad, I totally misunderstood. I add it then.

**Comment by GuillaumeGomez** [comment]
Updated.

File: src/librustc/diagnostics.rs [hunk 9]
```diff
@@ -360,6 +360,40 @@ integer type:
 http://doc.rust-lang.org/reference.html#ffi-attributes
 "##,
 
+E0109: r##"
+You tried to give type parameter to a type which doesn't need it. Erroneous
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: "a type parameter"

File: src/librustc/diagnostics.rs [hunk 10]
```diff
@@ -360,6 +360,40 @@ integer type:
...
+it doesn't need the type parameter.
+Example:
+
+```
+type X = u32; // ok!
+```
+"##,
+
+E0110: r##"
+You tried to give type parameter to a type which doesn't need it. Erroneous
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: "a lifetime parameter"

File: src/librustc/diagnostics.rs [hunk 11]
```diff
@@ -360,6 +360,40 @@ integer type:
...
+code example:
+
+```
+type X = u32<'static>; // error: lifetime parameters are not allowed on
+                       //        this type
+```
+
+Please check you actually used the good type or check again its definition.
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: Same corrections as the previous error.

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: On this one, I don't see the problem...

Comment (Manishearth) [inline src/librustc/diagnostics.rs]: "Please check that you used the correct type and recheck its definition, perhaps it doesn't need the lifetime parameter"

**Comment by GuillaumeGomez** [comment]
Updated ! Do you see anything else @Manishearth ?

File: src/librustc/diagnostics.rs [hunk 12]
```diff
@@ -360,6 +360,40 @@ integer type:
 http://doc.rust-lang.org/reference.html#ffi-attributes
 "##,
 
+E0109: r##"
+You tried to give    type parameter to a type which doesn't need it. Erroneous
```
Comment (Manishearth) [inline src/librustc/diagnostics.rs]: a

Comment (GuillaumeGomez) [inline src/librustc/diagnostics.rs]: The "a" has been eaten by the void ! :O

**Comment by GuillaumeGomez** [comment]
Re-updated.

**Pull request merged**